core/src/main/java/com/linecorp/armeria/server/logging/LoggingServiceBuilder.java

```diff
@@ -118,12 +128,57 @@ public LoggingServiceBuilder failureSamplingRate(float failureSamplingRate) {
         return failureSampler(Sampler.random(failureSamplingRate));
     }
 
+    /**
+     * Sets conditions to sample slow requests.
+     *
+     * <p>If {slowRequestPercentile} is 0.99, {windowMilliseconds} is 60000, we will sample requests that are
+     * slower than 99% of the requests within 1-minute time window. This will make sure our sampling adapts the
+     * traffic pattern. For example, traffic can be very low during the night thus our average response time
+     * will be pretty low. Therefore, our p99s should be lower than the midday traffic.
+     * Otherwise, we won't log any p99s during night.</p>
+     *
+     * <p>If we set {slowRequestSamplingLowerBoundMilliseconds} to 100, we won't sample any request that took
+     * less than 1000ms. This is useful to filter out requests that are too fast to be considered
+     * as slow requests. If your endpoint performs pretty healthy, you shouldn't see any slow request logs.</p>
+     *
+     * <p>If we set {slowRequestSamplingUpperBoundMilliseconds} to 1000, we will sample any request that
+     * took more than 1000 milliseconds. This is useful to make sure we log slow requests even if
+     * they are not in the p99 percentile. If your service is unhealthy, this will make sure any slow request
+     * is logged into your system.</p>
+     *
+     * @param slowRequestPercentile percentile of slow requests.
+     * @param windowMilliseconds window size to calculate percentile.
+     * @param slowRequestSamplingLowerBoundMilliseconds lower bound of slow requests.
+     *          Any request that took less than this amount of time won't be sampled.
+     * @param slowRequestSamplingUpperBoundMilliseconds upper bound of slow requests.
+     *          Any request that took more than this amount of time will be sampled regardless
+     *          of their percentile.
+     */
+    public LoggingServiceBuilder slowRequestSamplingPercentile(float slowRequestPercentile,
+                                                               long windowMilliseconds,
+                                                               long slowRequestSamplingLowerBoundMilliseconds,
+                                                               long slowRequestSamplingUpperBoundMilliseconds) {
+        final Sampler<Long> percentileMatches;
```

Comment:
Should we early return if some values are disabled and raise an exception if some of them are illegal?
```java
if ((slowRequestPercentile <= 0.0 || windowMilliseconds <= 0) &&
    slowRequestSamplingLowerBoundMilliseconds == 0 &&
    slowRequestSamplingUpperBoundMilliseconds == Long.MAX_VALUE) {
   return;
}
```

Reply:
Sure, I don't think I need to check the lower bound. Lower bound is just for ignoring so I guess users are free to pass there whatever they want regardless.